shepherd/shepherd/shepherd.py

```diff
@@ -266,6 +324,44 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
         except Exception:
             logging.exception('Error when reporting job `%s` as failed', job_id)
 
+    async def _finalize_job(self, message: Union[DoneMessage, ErrorMessage], sheep: BaseSheep) -> None:
+        """
+        Finalize a finished job - upload the results and update the status.
+
+        :param message: Message from the sheep about the end of the job
+        :param sheep: The sheep that sent the message
+        """
+
+        job_id = message.job_id
+
+        # clean-up the working directory and upload the results
+        working_directory = path.join(self._get_sheep(sheep.id).sheep_data_root, job_id)
+        await self._storage.push_job_data(job_id, working_directory)
+        shutil.rmtree(working_directory)
+
+        # save the done/error file
+        if isinstance(message, DoneMessage):
+            status = self._job_status.pop(job_id)
+            status.status = JobStatus.DONE
```

Comment:
I do not like how `DoneMessage` is processed in place while `ErrorMessage` is mostly handled with `self._report_job_failed`. I.e., we do quite the same at two different places and two different depths...